ICE: `Failed to normalize *mut std::pin::Pin<std::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()..`

<!--
[31mICE[0m: Rustc ./a.rs '-Zmir-opt-level=5 -Zcrate-attr=feature(async_drop) -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:171:13: Failed to normalize *mut std::pin::Pin<std::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), std::alloc::Global>> in typing_env=TypingEnv { typing_mode: PostAnalysis, param_env: ParamEnv { caller_bounds: [] } }, maybe try to call `try_normalize_erasing_regions` instead', 'error: internal compiler error: compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:171:13: Failed to normalize *mut std::pin::Pin<std::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), std::alloc::Global>> in typing_env=TypingEnv { typing_mode: PostAnalysis, param_env: ParamEnv { caller_bounds: [] } }, maybe try to call `try_normalize_erasing_regions` instead'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Zmir-opt-level=5
#![feature(async_drop)]
async fn call_once(f: impl AsyncFnOnce()) {
    let fut = Box::pin(f());
}

````

original:
````rust
//@ aux-build:block-on.rs
//@ edition:2018
//@ run-pass
//@ check-run-results

#![allow(unused)]

extern crate block_on;

struct DropMe(i32);

impl Drop for Drop {
    fn drop(&mut self) {
        println!("{} was dropped", self.0);
    }
}

async fn call_once(f: impl AsyncFnOnce()) {
    println!("before call");
    let fut = Box::pin(f());
    println!("after call");

    println!("future dropped");
}

fn main() {
    block_on::block_on(async {
        let d = DropMe(42);
        let async_closure = async move || {
            let d = &d;
            println!("called");
        };

        call_once(async_closure).await;
        println!("after");
    });
}

````

Version information
````
rustc 1.88.0-nightly (251cda5e1 2025-04-30)
binary: rustc
commit-hash: 251cda5e1f0057eb04fd9fc1653f2f1e010e8f97
commit-date: 2025-04-30
host: x86_64-unknown-linux-gnu
release: 1.88.0-nightly
LLVM version: 20.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/251cda5e1f0057eb04fd9fc1653f2f1e010e8f97/compiler/rustc_middle/src/ty/normalize_erasing_regions.rs#L165-L177

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zmir-opt-level=5 -Zcrate-attr=feature(async_drop)`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0670]: `async fn` is not permitted in Rust 2015
 --> /tmp/icemaker_global_tempdir.B6FTKk3Kw7JK/rustc_testrunner_tmpdir_reporting.mQsy3fxIN2EW/mvce.rs:1:1
  |
1 | async fn call_once(f: impl AsyncFnOnce()) {
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2024` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

warning: the feature `async_drop` is incomplete and may not be safe to use and/or cause compiler crashes
 --> <crate attribute>:1:12
  |
1 | #![feature(async_drop)]
  |            ^^^^^^^^^^
  |
  = note: see issue #126482 <https://github.com/rust-lang/rust/issues/126482> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.B6FTKk3Kw7JK/rustc_testrunner_tmpdir_reporting.mQsy3fxIN2EW/mvce.rs:3:2
  |
3 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.B6FTKk3Kw7JK/rustc_testrunner_tmpdir_reporting.mQsy3fxIN2EW/mvce.rs`

warning: unused variable: `fut`
 --> /tmp/icemaker_global_tempdir.B6FTKk3Kw7JK/rustc_testrunner_tmpdir_reporting.mQsy3fxIN2EW/mvce.rs:2:9
  |
2 |     let fut = Box::pin(f());
  |         ^^^ help: if this is intentional, prefix it with an underscore: `_fut`
  |
  = note: `#[warn(unused_variables)]` on by default

error: internal compiler error: compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:171:13: Failed to normalize *mut std::pin::Pin<std::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), std::alloc::Global>> in typing_env=TypingEnv { typing_mode: PostAnalysis, param_env: ParamEnv { caller_bounds: [] } }, maybe try to call `try_normalize_erasing_regions` instead


thread 'rustc' panicked at compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:171:13:
Box<dyn Any>
stack backtrace:
   0:     0x7ba035563863 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h20b93d714a840c17
   1:     0x7ba035c059c7 - core::fmt::write::hbd4e43071f2b6280
   2:     0x7ba03700a091 - std::io::Write::write_fmt::hd6609147c061a65a
   3:     0x7ba0355636c2 - std::sys::backtrace::BacktraceLock::print::he17c913a476f0df2
   4:     0x7ba03556750a - std::panicking::default_hook::{{closure}}::h387e3e0861f6ca39
   5:     0x7ba03556708f - std::panicking::default_hook::h06a982276987baa3
   6:     0x7ba0345a7633 - std[553e8cb93e7fb90]::panicking::update_hook::<alloc[6998e39f333261e4]::boxed::Box<rustc_driver_impl[48b038e11ca70c04]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7ba035567d83 - std::panicking::rust_panic_with_hook::h0279421d7919da67
   8:     0x7ba0345e31a1 - std[553e8cb93e7fb90]::panicking::begin_panic::<rustc_errors[9c6bd54a8b3d7a47]::ExplicitBug>::{closure#0}
   9:     0x7ba0345d71d6 - std[553e8cb93e7fb90]::sys::backtrace::__rust_end_short_backtrace::<std[553e8cb93e7fb90]::panicking::begin_panic<rustc_errors[9c6bd54a8b3d7a47]::ExplicitBug>::{closure#0}, !>
  10:     0x7ba0345d7063 - std[553e8cb93e7fb90]::panicking::begin_panic::<rustc_errors[9c6bd54a8b3d7a47]::ExplicitBug>
  11:     0x7ba0345ed811 - <rustc_errors[9c6bd54a8b3d7a47]::diagnostic::BugAbort as rustc_errors[9c6bd54a8b3d7a47]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7ba034c48f1a - rustc_middle[fe83ff8f6af0fa19]::util::bug::opt_span_bug_fmt::<rustc_span[1adb01313f1c57dc]::span_encoding::Span>::{closure#0}
  13:     0x7ba034c2177a - rustc_middle[fe83ff8f6af0fa19]::ty::context::tls::with_opt::<rustc_middle[fe83ff8f6af0fa19]::util::bug::opt_span_bug_fmt<rustc_span[1adb01313f1c57dc]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:     0x7ba034c215eb - rustc_middle[fe83ff8f6af0fa19]::ty::context::tls::with_context_opt::<rustc_middle[fe83ff8f6af0fa19]::ty::context::tls::with_opt<rustc_middle[fe83ff8f6af0fa19]::util::bug::opt_span_bug_fmt<rustc_span[1adb01313f1c57dc]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:     0x7ba031e9d630 - rustc_middle[fe83ff8f6af0fa19]::util::bug::bug_fmt
  16:     0x7ba03659c942 - <rustc_middle[fe83ff8f6af0fa19]::ty::normalize_erasing_regions::NormalizeAfterErasingRegionsFolder as rustc_type_ir[72209fb0c9973b28]::fold::TypeFolder<rustc_middle[fe83ff8f6af0fa19]::ty::context::TyCtxt>>::fold_ty
  17:     0x7ba036f6967d - rustc_mir_transform[13028d53c6d11955]::validate::validate_types
  18:     0x7ba036cf49fe - <rustc_mir_transform[13028d53c6d11955]::validate::Validator as rustc_mir_transform[13028d53c6d11955]::pass_manager::MirPass>::run_pass
  19:     0x7ba035c03c4e - rustc_mir_transform[13028d53c6d11955]::pass_manager::run_passes_inner
  20:     0x7ba0365c707d - rustc_mir_transform[13028d53c6d11955]::run_optimization_passes
  21:     0x7ba03641e67e - rustc_mir_transform[13028d53c6d11955]::shim::make_shim
  22:     0x7ba03641c62f - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::mir_shims::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 8usize]>>
  23:     0x7ba03641c5e7 - <rustc_query_impl[73ffed637ff28f75]::query_impl::mir_shims::dynamic_query::{closure#2} as core[f0a60fd3088b4ecc]::ops::function::FnOnce<(rustc_middle[fe83ff8f6af0fa19]::ty::context::TyCtxt, rustc_middle[fe83ff8f6af0fa19]::ty::instance::InstanceKind)>>::call_once
  24:     0x7ba03631a8d2 - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::instance::InstanceKind, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  25:     0x7ba03631a5eb - rustc_query_impl[73ffed637ff28f75]::query_impl::mir_shims::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7ba034c45d80 - <rustc_middle[fe83ff8f6af0fa19]::ty::context::TyCtxt>::coroutine_layout
  27:     0x7ba0362281cf - rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached
  28:     0x7ba036216e4e - rustc_ty_utils[7e79272ab878db29]::layout::layout_of
  29:     0x7ba036216db2 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>
  30:     0x7ba036215baf - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::PseudoCanonicalInput<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  31:     0x7ba0362157ee - rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  32:     0x7ba0362191a0 - <core[f0a60fd3088b4ecc]::iter::adapters::GenericShunt<core[f0a60fd3088b4ecc]::iter::adapters::by_ref_sized::ByRefSized<core[f0a60fd3088b4ecc]::iter::adapters::map::Map<core[f0a60fd3088b4ecc]::slice::iter::Iter<rustc_middle[fe83ff8f6af0fa19]::ty::VariantDef>, rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached::{closure#20}>>, core[f0a60fd3088b4ecc]::result::Result<core[f0a60fd3088b4ecc]::convert::Infallible, &rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutError>> as core[f0a60fd3088b4ecc]::iter::traits::iterator::Iterator>::next
  33:     0x7ba03621d244 - rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached
  34:     0x7ba036216e4e - rustc_ty_utils[7e79272ab878db29]::layout::layout_of
  35:     0x7ba036216db2 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>
  36:     0x7ba036215baf - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::PseudoCanonicalInput<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  37:     0x7ba0362157ee - rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7ba036219550 - <core[f0a60fd3088b4ecc]::iter::adapters::GenericShunt<core[f0a60fd3088b4ecc]::iter::adapters::by_ref_sized::ByRefSized<core[f0a60fd3088b4ecc]::iter::adapters::map::Map<core[f0a60fd3088b4ecc]::slice::iter::Iter<rustc_middle[fe83ff8f6af0fa19]::ty::VariantDef>, rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached::{closure#20}>>, core[f0a60fd3088b4ecc]::result::Result<core[f0a60fd3088b4ecc]::convert::Infallible, &rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutError>> as core[f0a60fd3088b4ecc]::iter::traits::iterator::Iterator>::next
  39:     0x7ba03621d244 - rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached
  40:     0x7ba036216e4e - rustc_ty_utils[7e79272ab878db29]::layout::layout_of
  41:     0x7ba036216db2 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>
  42:     0x7ba036215baf - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::PseudoCanonicalInput<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  43:     0x7ba0362157ee - rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  44:     0x7ba0362178a8 - rustc_ty_utils[7e79272ab878db29]::layout::layout_of
  45:     0x7ba036216db2 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>
  46:     0x7ba036215baf - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::PseudoCanonicalInput<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  47:     0x7ba0362157ee - rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  48:     0x7ba036214d89 - <rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutCx as rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutOf>::spanned_layout_of
  49:     0x7ba0354cf3b0 - <core[f0a60fd3088b4ecc]::iter::adapters::GenericShunt<core[f0a60fd3088b4ecc]::iter::adapters::by_ref_sized::ByRefSized<core[f0a60fd3088b4ecc]::iter::adapters::map::Map<core[f0a60fd3088b4ecc]::slice::iter::Iter<rustc_middle[fe83ff8f6af0fa19]::mir::query::CoroutineSavedTy>, rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached::{closure#12}>>, core[f0a60fd3088b4ecc]::result::Result<core[f0a60fd3088b4ecc]::convert::Infallible, &rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutError>> as core[f0a60fd3088b4ecc]::iter::traits::iterator::Iterator>::next
  50:     0x7ba0354babda - core[f0a60fd3088b4ecc]::iter::adapters::try_process::<core[f0a60fd3088b4ecc]::iter::adapters::by_ref_sized::ByRefSized<core[f0a60fd3088b4ecc]::iter::adapters::map::Map<core[f0a60fd3088b4ecc]::slice::iter::Iter<rustc_middle[fe83ff8f6af0fa19]::mir::query::CoroutineSavedTy>, rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached::{closure#12}>>, rustc_abi[c43837ab31a6f7c2]::layout::ty::TyAndLayout<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, core[f0a60fd3088b4ecc]::result::Result<core[f0a60fd3088b4ecc]::convert::Infallible, &rustc_middle[fe83ff8f6af0fa19]::ty::layout::LayoutError>, <core[f0a60fd3088b4ecc]::iter::adapters::map::Map<core[f0a60fd3088b4ecc]::slice::iter::Iter<rustc_middle[fe83ff8f6af0fa19]::mir::query::CoroutineSavedTy>, rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached::{closure#12}> as core[f0a60fd3088b4ecc]::iter::traits::iterator::Iterator>::try_collect<rustc_index[9438a2ca845a1647]::vec::IndexVec<rustc_middle[fe83ff8f6af0fa19]::mir::query::CoroutineSavedLocal, rustc_abi[c43837ab31a6f7c2]::layout::ty::TyAndLayout<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>>>::{closure#0}, rustc_index[9438a2ca845a1647]::vec::IndexVec<rustc_middle[fe83ff8f6af0fa19]::mir::query::CoroutineSavedLocal, rustc_abi[c43837ab31a6f7c2]::layout::ty::TyAndLayout<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>>>
  51:     0x7ba0362287ef - rustc_ty_utils[7e79272ab878db29]::layout::layout_of_uncached
  52:     0x7ba036216e4e - rustc_ty_utils[7e79272ab878db29]::layout::layout_of
  53:     0x7ba036216db2 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>
  54:     0x7ba036215baf - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::DefaultCache<rustc_middle[fe83ff8f6af0fa19]::ty::PseudoCanonicalInput<rustc_middle[fe83ff8f6af0fa19]::ty::Ty>, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  55:     0x7ba0362157ee - rustc_query_impl[73ffed637ff28f75]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  56:     0x7ba036a5c9ee - <rustc_mir_transform[13028d53c6d11955]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[13028d53c6d11955]::pass_manager::MirLint>::run_lint
  57:     0x7ba035c0bea7 - rustc_mir_transform[13028d53c6d11955]::run_analysis_to_runtime_passes
  58:     0x7ba03645abe5 - rustc_mir_transform[13028d53c6d11955]::mir_drops_elaborated_and_const_checked
  59:     0x7ba03645a5b7 - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 8usize]>>
  60:     0x7ba035f8746c - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_data_structures[cad23842afe3738d]::vec_cache::VecCache<rustc_span[1adb01313f1c57dc]::def_id::LocalDefId, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[341c3edc9f45c813]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  61:     0x7ba035f86d4f - rustc_query_impl[73ffed637ff28f75]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  62:     0x7ba0366e66d8 - rustc_interface[39a942c1d345d61e]::passes::run_required_analyses
  63:     0x7ba036af739e - rustc_interface[39a942c1d345d61e]::passes::analysis
  64:     0x7ba036af736d - rustc_query_impl[73ffed637ff28f75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[73ffed637ff28f75]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 0usize]>>
  65:     0x7ba036afce7d - rustc_query_system[341c3edc9f45c813]::query::plumbing::try_execute_query::<rustc_query_impl[73ffed637ff28f75]::DynamicConfig<rustc_query_system[341c3edc9f45c813]::query::caches::SingleCache<rustc_middle[fe83ff8f6af0fa19]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[73ffed637ff28f75]::plumbing::QueryCtxt, false>
  66:     0x7ba036afcb78 - rustc_query_impl[73ffed637ff28f75]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  67:     0x7ba036b3a8be - rustc_interface[39a942c1d345d61e]::passes::create_and_enter_global_ctxt::<core[f0a60fd3088b4ecc]::option::Option<rustc_interface[39a942c1d345d61e]::queries::Linker>, rustc_driver_impl[48b038e11ca70c04]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  68:     0x7ba036c6b254 - rustc_interface[39a942c1d345d61e]::interface::run_compiler::<(), rustc_driver_impl[48b038e11ca70c04]::run_compiler::{closure#0}>::{closure#1}
  69:     0x7ba036c92c74 - std[553e8cb93e7fb90]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[39a942c1d345d61e]::util::run_in_thread_with_globals<rustc_interface[39a942c1d345d61e]::util::run_in_thread_pool_with_globals<rustc_interface[39a942c1d345d61e]::interface::run_compiler<(), rustc_driver_impl[48b038e11ca70c04]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  70:     0x7ba036c930b4 - <<std[553e8cb93e7fb90]::thread::Builder>::spawn_unchecked_<rustc_interface[39a942c1d345d61e]::util::run_in_thread_with_globals<rustc_interface[39a942c1d345d61e]::util::run_in_thread_pool_with_globals<rustc_interface[39a942c1d345d61e]::interface::run_compiler<(), rustc_driver_impl[48b038e11ca70c04]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[f0a60fd3088b4ecc]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  71:     0x7ba036c944ab - std::sys::pal::unix::thread::Thread::new::thread_start::h263b7b8b08c5ad3d
  72:     0x7ba030aa370a - <unknown>
  73:     0x7ba030b27aac - <unknown>
  74:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.88.0-nightly (251cda5e1 2025-04-30) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=5 -Z crate-attr=feature(async_drop) -Z dump-mir-dir=dir

query stack during panic:
#0 [mir_shims] generating MIR shim for `core::future::async_drop::async_drop_in_place::{closure#0}`, instance=AsyncDropGlue(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), Coroutine(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), [core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>, (), core::future::ResumeTy, (), (), CoroutineWitness(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), [core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>]), (*mut core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>,)]))
#1 [layout_of] computing layout of `{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}`
#2 [layout_of] computing layout of `core::mem::manually_drop::ManuallyDrop<{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}>`
#3 [layout_of] computing layout of `core::mem::maybe_uninit::MaybeUninit<{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}>`
#4 [layout_of] computing layout of `core::mem::maybe_uninit::MaybeUninit<core::future::async_drop::async_drop_in_place::{opaque#0}>`
#5 [layout_of] computing layout of `{async fn body of call_once<impl AsyncFnOnce()>()}`
#6 [mir_drops_elaborated_and_const_checked] elaborating drops for `call_once`
#7 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0601, E0670.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
error: internal compiler error: compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:171:13: Failed to normalize *mut std::pin::Pin<std::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), std::alloc::Global>> in typing_env=TypingEnv { typing_mode: PostAnalysis, param_env: ParamEnv { caller_bounds: [] } }, maybe try to call `try_normalize_erasing_regions` instead
#0 [mir_shims] generating MIR shim for `core::future::async_drop::async_drop_in_place::{closure#0}`, instance=AsyncDropGlue(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), Coroutine(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), [core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>, (), core::future::ResumeTy, (), (), CoroutineWitness(DefId(2:15313 ~ core[f0a6]::future::async_drop::async_drop_in_place::{closure#0}), [core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>]), (*mut core::pin::Pin<alloc::boxed::Box<Alias(Projection, AliasTy { args: [impl AsyncFnOnce()/#0, ()], def_id: DefId(2:3751 ~ core[f0a6]::ops::async_function::AsyncFnOnce::CallOnceFuture), .. }), alloc::alloc::Global>>,)]))
#1 [layout_of] computing layout of `{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}`
#2 [layout_of] computing layout of `core::mem::manually_drop::ManuallyDrop<{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}>`
#3 [layout_of] computing layout of `core::mem::maybe_uninit::MaybeUninit<{async fn body of core::future::async_drop::async_drop_in_place<core::pin::Pin<alloc::boxed::Box<<impl AsyncFnOnce() as core::ops::async_function::AsyncFnOnce<()>>::CallOnceFuture>>>()}>`
#4 [layout_of] computing layout of `core::mem::maybe_uninit::MaybeUninit<core::future::async_drop::async_drop_in_place::{opaque#0}>`
#5 [layout_of] computing layout of `{async fn body of call_once<impl AsyncFnOnce()>()}`
#6 [mir_drops_elaborated_and_const_checked] elaborating drops for `call_once`
#7 [analysis] running analysis passes on this crate

-->
@rustbot label +F-async_drop

--- a/compiler/rustc_mir_transform/src/coroutine/drop.rs
+++ b/compiler/rustc_mir_transform/src/coroutine/drop.rs
@@ -132,6 +132,7 @@ fn build_poll_switch<'tcx>(
     body: &mut Body<'tcx>,
     poll_enum: Ty<'tcx>,
     poll_unit_place: &Place<'tcx>,
+    fut_pin_place: &Place<'tcx>,
     ready_block: BasicBlock,
     yield_block: BasicBlock,
 ) -> BasicBlock {
@@ -162,9 +163,11 @@ fn build_poll_switch<'tcx>(
             Rvalue::Discriminant(*poll_unit_place),
         ))),
     };
+    let storage_dead =
+        Statement { source_info, kind: StatementKind::StorageDead(fut_pin_place.local) };
     let unreachable_block = insert_term_block(body, TerminatorKind::Unreachable);
     body.basic_blocks_mut().push(BasicBlockData {
-        statements: [discr_assign].to_vec(),
+        statements: [storage_dead, discr_assign].to_vec(),
         terminator: Some(Terminator {
             source_info,
             kind: TerminatorKind::SwitchInt {
@@ -332,10 +335,17 @@ pub(super) fn expand_async_drops<'tcx>(
             kind: StatementKind::Assign(Box::new((context_ref_place, arg))),
         });
         let yield_block = insert_term_block(body, TerminatorKind::Unreachable); // `kind` replaced later to yield
-        let switch_block =
-            build_poll_switch(tcx, body, poll_enum, &poll_unit_place, target, yield_block);
         let (pin_bb, fut_pin_place) =
             build_pin_fut(tcx, body, fut_place.clone(), UnwindAction::Continue);
+        let switch_block = build_poll_switch(
+            tcx,
+            body,
+            poll_enum,
+            &poll_unit_place,
+            &fut_pin_place,
+            target,
+            yield_block,
+        );
         let call_bb = build_poll_call(
             tcx,
             body,
@@ -357,16 +367,17 @@ pub(super) fn expand_async_drops<'tcx>(
                 body.local_decls.push(LocalDecl::new(context_mut_ref, source_info.span)),
             );
             let drop_yield_block = insert_term_block(body, TerminatorKind::Unreachable); // `kind` replaced later to yield
+            let (pin_bb2, fut_pin_place2) =
+                build_pin_fut(tcx, body, fut_place, UnwindAction::Continue);
             let drop_switch_block = build_poll_switch(
                 tcx,
                 body,
                 poll_enum,
                 &poll_unit_place,
+                &fut_pin_place2,
                 drop.unwrap(),
                 drop_yield_block,
             );
-            let (pin_bb2, fut_pin_place2) =
-                build_pin_fut(tcx, body, fut_place, UnwindAction::Continue);
             let drop_call_bb = build_poll_call(
                 tcx,
                 body,

--- a/compiler/rustc_mir_transform/src/elaborate_drop.rs
+++ b/compiler/rustc_mir_transform/src/elaborate_drop.rs
@@ -390,6 +390,20 @@ where
             Location { block: self.succ, statement_index: 0 },
             StatementKind::StorageDead(fut.local),
         );
+        // StorageDead(fut) in unwind block (at the begin)
+        if let Unwind::To(block) = unwind {
+            self.elaborator.patch().add_statement(
+                Location { block, statement_index: 0 },
+                StatementKind::StorageDead(fut.local),
+            );
+        }
+        // StorageDead(fut) in dropline block (at the begin)
+        if let Some(block) = dropline {
+            self.elaborator.patch().add_statement(
+                Location { block, statement_index: 0 },
+                StatementKind::StorageDead(fut.local),
+            );
+        }
 
         // #1:pin_obj_bb >>> call Pin<ObjTy>::new_unchecked(&mut obj)
         self.elaborator.patch().patch_terminator(

--- a/tests/mir-opt/async_drop_live_dead.a-{closure#0}.coroutine_drop_async.0.panic-abort.mir
+++ b/tests/mir-opt/async_drop_live_dead.a-{closure#0}.coroutine_drop_async.0.panic-abort.mir
@@ -1,0 +1,100 @@
+// MIR for `a::{closure#0}` 0 coroutine_drop_async
+
+fn a::{closure#0}(_1: Pin<&mut {async fn body of a<T>()}>, _2: &mut Context<'_>) -> Poll<()> {
+    debug _task_context => _19;
+    debug x => ((*(_1.0: &mut {async fn body of a<T>()})).0: T);
+    let mut _0: std::task::Poll<()>;
+    let _3: T;
+    let mut _4: impl std::future::Future<Output = ()>;
+    let mut _5: &mut T;
+    let mut _6: std::pin::Pin<&mut T>;
+    let mut _7: &mut T;
+    let mut _8: *mut T;
+    let mut _9: ();
+    let mut _10: std::task::Poll<()>;
+    let mut _11: &mut std::task::Context<'_>;
+    let mut _12: &mut impl std::future::Future<Output = ()>;
+    let mut _13: std::pin::Pin<&mut impl std::future::Future<Output = ()>>;
+    let mut _14: isize;
+    let mut _15: &mut std::task::Context<'_>;
+    let mut _16: &mut impl std::future::Future<Output = ()>;
+    let mut _17: std::pin::Pin<&mut impl std::future::Future<Output = ()>>;
+    let mut _18: isize;
+    let mut _19: &mut std::task::Context<'_>;
+    let mut _20: u32;
+    scope 1 {
+        debug x => (((*(_1.0: &mut {async fn body of a<T>()})) as variant#4).0: T);
+    }
+
+    bb0: {
+        _20 = discriminant((*(_1.0: &mut {async fn body of a<T>()})));
+        switchInt(move _20) -> [0: bb9, 3: bb12, 4: bb13, otherwise: bb14];
+    }
+
+    bb1: {
+        nop;
+        nop;
+        goto -> bb2;
+    }
+
+    bb2: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+
+    bb3: {
+        _0 = Poll::<()>::Pending;
+        discriminant((*(_1.0: &mut {async fn body of a<T>()}))) = 4;
+        return;
+    }
+
+    bb4: {
+        StorageLive(_17);
+        _16 = &mut (((*(_1.0: &mut {async fn body of a<T>()})) as variant#4).1: impl std::future::Future<Output = ()>);
+        _17 = Pin::<&mut impl Future<Output = ()>>::new_unchecked(move _16) -> [return: bb7, unwind unreachable];
+    }
+
+    bb5: {
+        unreachable;
+    }
+
+    bb6: {
+        StorageDead(_17);
+        _18 = discriminant(_10);
+        switchInt(move _18) -> [0: bb1, 1: bb3, otherwise: bb5];
+    }
+
+    bb7: {
+        _10 = <impl Future<Output = ()> as Future>::poll(move _17, move _15) -> [return: bb6, unwind unreachable];
+    }
+
+    bb8: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+
+    bb9: {
+        goto -> bb11;
+    }
+
+    bb10: {
+        goto -> bb8;
+    }
+
+    bb11: {
+        drop(((*(_1.0: &mut {async fn body of a<T>()})).0: T)) -> [return: bb10, unwind unreachable];
+    }
+
+    bb12: {
+        goto -> bb4;
+    }
+
+    bb13: {
+        goto -> bb4;
+    }
+
+    bb14: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+}

--- a/tests/mir-opt/async_drop_live_dead.a-{closure#0}.coroutine_drop_async.0.panic-unwind.mir
+++ b/tests/mir-opt/async_drop_live_dead.a-{closure#0}.coroutine_drop_async.0.panic-unwind.mir
@@ -1,0 +1,123 @@
+// MIR for `a::{closure#0}` 0 coroutine_drop_async
+
+fn a::{closure#0}(_1: Pin<&mut {async fn body of a<T>()}>, _2: &mut Context<'_>) -> Poll<()> {
+    debug _task_context => _19;
+    debug x => ((*(_1.0: &mut {async fn body of a<T>()})).0: T);
+    let mut _0: std::task::Poll<()>;
+    let _3: T;
+    let mut _4: impl std::future::Future<Output = ()>;
+    let mut _5: &mut T;
+    let mut _6: std::pin::Pin<&mut T>;
+    let mut _7: &mut T;
+    let mut _8: *mut T;
+    let mut _9: ();
+    let mut _10: std::task::Poll<()>;
+    let mut _11: &mut std::task::Context<'_>;
+    let mut _12: &mut impl std::future::Future<Output = ()>;
+    let mut _13: std::pin::Pin<&mut impl std::future::Future<Output = ()>>;
+    let mut _14: isize;
+    let mut _15: &mut std::task::Context<'_>;
+    let mut _16: &mut impl std::future::Future<Output = ()>;
+    let mut _17: std::pin::Pin<&mut impl std::future::Future<Output = ()>>;
+    let mut _18: isize;
+    let mut _19: &mut std::task::Context<'_>;
+    let mut _20: u32;
+    scope 1 {
+        debug x => (((*(_1.0: &mut {async fn body of a<T>()})) as variant#4).0: T);
+    }
+
+    bb0: {
+        _20 = discriminant((*(_1.0: &mut {async fn body of a<T>()})));
+        switchInt(move _20) -> [0: bb12, 2: bb18, 3: bb16, 4: bb17, otherwise: bb19];
+    }
+
+    bb1: {
+        nop;
+        nop;
+        goto -> bb2;
+    }
+
+    bb2: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+
+    bb3 (cleanup): {
+        nop;
+        nop;
+        goto -> bb5;
+    }
+
+    bb4 (cleanup): {
+        goto -> bb15;
+    }
+
+    bb5 (cleanup): {
+        goto -> bb4;
+    }
+
+    bb6: {
+        _0 = Poll::<()>::Pending;
+        discriminant((*(_1.0: &mut {async fn body of a<T>()}))) = 4;
+        return;
+    }
+
+    bb7: {
+        StorageLive(_17);
+        _16 = &mut (((*(_1.0: &mut {async fn body of a<T>()})) as variant#4).1: impl std::future::Future<Output = ()>);
+        _17 = Pin::<&mut impl Future<Output = ()>>::new_unchecked(move _16) -> [return: bb10, unwind: bb15];
+    }
+
+    bb8: {
+        unreachable;
+    }
+
+    bb9: {
+        StorageDead(_17);
+        _18 = discriminant(_10);
+        switchInt(move _18) -> [0: bb1, 1: bb6, otherwise: bb8];
+    }
+
+    bb10: {
+        _10 = <impl Future<Output = ()> as Future>::poll(move _17, move _15) -> [return: bb9, unwind: bb3];
+    }
+
+    bb11: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+
+    bb12: {
+        goto -> bb14;
+    }
+
+    bb13: {
+        goto -> bb11;
+    }
+
+    bb14: {
+        drop(((*(_1.0: &mut {async fn body of a<T>()})).0: T)) -> [return: bb13, unwind: bb4];
+    }
+
+    bb15 (cleanup): {
+        discriminant((*(_1.0: &mut {async fn body of a<T>()}))) = 2;
+        resume;
+    }
+
+    bb16: {
+        goto -> bb7;
+    }
+
+    bb17: {
+        goto -> bb7;
+    }
+
+    bb18: {
+        assert(const false, "`async fn` resumed after panicking") -> [success: bb18, unwind continue];
+    }
+
+    bb19: {
+        _0 = Poll::<()>::Ready(const ());
+        return;
+    }
+}

--- a/tests/mir-opt/async_drop_live_dead.rs
+++ b/tests/mir-opt/async_drop_live_dead.rs
@@ -1,0 +1,11 @@
+//@ edition:2024
+// skip-filecheck
+// EMIT_MIR_FOR_EACH_PANIC_STRATEGY
+
+#![feature(async_drop)]
+#![allow(incomplete_features)]
+
+// EMIT_MIR async_drop_live_dead.a-{closure#0}.coroutine_drop_async.0.mir
+async fn a<T>(x: T) {}
+
+fn main() {}

--- a/tests/ui/async-await/async-drop/async-drop-initial.rs
+++ b/tests/ui/async-await/async-drop/async-drop-initial.rs
@@ -62,7 +62,7 @@ fn main() {
         test_async_drop(&j, 16).await;
         test_async_drop(
             AsyncStruct { b: AsyncInt(8), a: AsyncInt(7), i: 6 },
-            if cfg!(panic = "unwind") { 168 } else { 136 },
+            136,
         ).await;
         test_async_drop(ManuallyDrop::new(AsyncInt(9)), 16).await;
 

--- a/tests/ui/async-await/async-drop/live-dead-storage.rs
+++ b/tests/ui/async-await/async-drop/live-dead-storage.rs
@@ -1,6 +1,9 @@
-//@ known-bug: #140429
+// ex-ice: #140429
 //@ compile-flags: -Zlint-mir --crate-type lib
 //@ edition:2024
+//@ check-pass
 
 #![feature(async_drop)]
+#![allow(incomplete_features)]
+
 async fn a<T>(x: T) {}

--- a/tests/ui/async-await/async-drop/live-dead-storage2.rs
+++ b/tests/ui/async-await/async-drop/live-dead-storage2.rs
@@ -1,7 +1,11 @@
-//@ known-bug: #140531
-//@compile-flags: -Zlint-mir --crate-type lib
+// ex-ice: #140531
+//@ compile-flags: -Zlint-mir --crate-type lib
 //@ edition:2024
+//@ check-pass
+
 #![feature(async_drop)]
+#![allow(incomplete_features)]
+
 async fn call_once(f: impl AsyncFnOnce()) {
     let fut = Box::pin(f());
 }

--- a/tests/ui/async-await/async-drop/live-dead-storage3.rs
+++ b/tests/ui/async-await/async-drop/live-dead-storage3.rs
@@ -1,0 +1,56 @@
+// ex-ice: #141761
+//@ compile-flags: -Zlint-mir --crate-type lib
+//@ edition:2024
+//@ check-pass
+
+#![feature(async_drop)]
+#![allow(incomplete_features)]
+
+type BoxFuture<T> = std::pin::Pin<Box<dyn Future<Output = T>>>;
+fn main() {}
+async fn f() {
+    run("").await
+}
+struct InMemoryStorage;
+struct User<'dep> {
+    dep: &'dep str,
+}
+impl<'a> StorageRequest<InMemoryStorage> for SaveUser<'a> {
+    fn execute(&self) -> BoxFuture<Result<(), String>> {
+        todo!()
+    }
+}
+trait Storage {
+    type Error;
+}
+impl Storage for InMemoryStorage {
+    type Error = String;
+}
+trait StorageRequestReturnType {
+    type Output;
+}
+trait StorageRequest<S: Storage>: StorageRequestReturnType {
+    fn execute(&self) -> BoxFuture<Result<<Self>::Output, S::Error>>;
+}
+struct SaveUser<'a> {
+    name: &'a str,
+}
+impl<'a> StorageRequestReturnType for SaveUser<'a> {
+    type Output = ();
+}
+impl<'dep> User<'dep> {
+    async fn save<S>(self)
+    where
+        S: Storage,
+        for<'a> SaveUser<'a>: StorageRequest<S>,
+    {
+        SaveUser { name: "" }.execute().await;
+    }
+}
+async fn run<S>(dep: &str)
+where
+    S: Storage,
+    for<'a> SaveUser<'a>: StorageRequest<S>,
+{
+    User { dep }.save().await
+}

--- a/tests/ui/async-await/async-drop/live-dead-storage4.rs
+++ b/tests/ui/async-await/async-drop/live-dead-storage4.rs
@@ -1,0 +1,56 @@
+// ex-ice: #141409
+//@ compile-flags: -Zmir-enable-passes=+Inline -Zvalidate-mir -Zlint-mir --crate-type lib
+//@ edition:2024
+//@ check-pass
+
+#![feature(async_drop)]
+#![allow(incomplete_features)]
+#![allow(non_snake_case)]
+
+use std::mem::ManuallyDrop;
+use std::{
+    future::{async_drop_in_place, Future},
+    pin::{pin, Pin},
+    sync::{mpsc, Arc},
+    task::{Context, Poll, Wake, Waker},
+};
+fn main() {
+    block_on(bar(0))
+}
+async fn baz(ident_base: usize) {}
+async fn bar(ident_base: usize) {
+    baz(1).await
+}
+fn block_on<F>(fut_unpin: F) -> F::Output
+where
+    F: Future,
+{
+    let fut_pin = pin!(ManuallyDrop::new(fut_unpin));
+    let mut fut = unsafe { Pin::map_unchecked_mut(fut_pin, |x| &mut **x) };
+    let (waker, rx) = simple_waker();
+    let mut context = Context::from_waker(&waker);
+    let rv = loop {
+        match fut.as_mut().poll(&mut context) {
+            Poll::Ready(out) => break out,
+            PollPending => (),
+        }
+    };
+    let drop_fut_unpin = unsafe { async_drop_in_place(fut.get_unchecked_mut()) };
+    let drop_fut = pin!(drop_fut_unpin);
+    loop {
+        match drop_fut.poll(&mut context) {
+            Poll => break,
+        }
+    }
+    rv
+}
+fn simple_waker() -> (Waker, mpsc::Receiver<()>) {
+    struct SimpleWaker {
+        tx: mpsc::Sender<()>,
+    }
+    impl Wake for SimpleWaker {
+        fn wake(self: Arc<Self>) {}
+    }
+    let (tx, rx) = mpsc::channel();
+    (Waker::from(Arc::new(SimpleWaker { tx })), rx)
+}

